### PR TITLE
chore(eslint-plugin): refactor explicit return type rules to share code

### PR DESCRIPTION
**Pull request by bradzacher**
I saw this duplication, but let it go in the PR because it was a non-trivial refactor.
This PR just cleans up that tech debt.

### CODE REVIEW

File: packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+/**
+ * Checks if a node belongs to:
+ * ```
+ * new Foo(() => {})
+ *         ^^^^^^^^
+ * ```
+ */
+function isConstructorArgument(parent: TSESTree.Node): boolean {
```
Comment (armano2) [inline packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts]:
```suggestion
function isConstructorArgument(parent: TSESTree.Node): parent is TSESTree.NewExpression {
```

File: packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts [hunk 2]
```diff
@@ -88,7 +90,9 @@ function isVariableDeclaratorWithTypeAnnotation(node: TSESTree.Node): boolean {
  * public x: Foo = ...
  * ```
  */
-function isClassPropertyWithTypeAnnotation(node: TSESTree.Node): boolean {
+function isClassPropertyWithTypeAnnotation(
+  node: TSESTree.Node,
+): node is TSESTree.ClassProperty {
```
Comment (armano2) [inline packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts]: this is not correct, negated types does not include `ClassProperty`

File: packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts [hunk 3]
```diff
@@ -76,7 +76,9 @@ function getReporLoc(
  * const x: Foo = ...
  * ```
  */
-function isVariableDeclaratorWithTypeAnnotation(node: TSESTree.Node): boolean {
+function isVariableDeclaratorWithTypeAnnotation(
+  node: TSESTree.Node,
+): node is TSESTree.VariableDeclarator {
```
Comment (armano2) [inline packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts]: this is not correct, negated types does not include `VariableDeclarator`

Comment (bradzacher) [reply packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts]: what do you mean by this, sorry?

The function checks `node.type === AST_NODE_TYPES.VariableDeclarator`, so this is the correct guard.
I just didn't bother doing the `& { id: TSESTree.Identifier & { typeAnnotation: TSESTree.Identifier['typeAnnotation'] }`

**Pull request merged**